src/groupeddataframe/grouping.jl

```diff
@@ -177,6 +213,292 @@ end
+    end
+end
+
+function Base.NamedTuple(key::GroupKey)
+    N = NamedTuple{Tuple(groupvars(parent(key)))}
+    N(_groupvalues(parent(key), getfield(key, :idx)))
+end
+Base.Tuple(key::GroupKey) = values(key)
+
+
+"""
+    GroupKeys{T<:GroupedDataFrame} <: AbstractVector{GroupKey{T}}
+
+An abstract vector containing all [`GroupKey`](@ref) objects for a given
+[`GroupedDataFrame`](@ref).
+
+See [`keys(::GroupedDataFrame)`](@ref) for more information.
+"""
+struct GroupKeys{T<:GroupedDataFrame} <: AbstractVector{GroupKey{T}}
+    parent::T
+end
+
+Base.parent(gk::GroupKeys) = gk.parent
+
+Base.size(gk::GroupKeys) = (length(parent(gk)),)
+Base.IndexStyle(::GroupKeys) = IndexLinear()
+function Base.getindex(gk::GroupKeys, i::Integer)
+    @boundscheck checkbounds(gk, i)
+    return GroupKey(parent(gk), i)
+end
+
+
+"""
+    keys(gd::GroupedDataFrame)
+
+Get the set of keys for each group of the `GroupedDataFrame` `gd` as a
+[`GroupKeys`](@ref) object. Each key is a [`GroupKey`](@ref), which behaves like
+a `NamedTuple` holding the values of the grouping columns for a given group.
+Unlike the equivalent `Tuple` and `NamedTuple`, these keys can be used to index
+into `gd` efficiently. The ordering of the keys is identical to the ordering of
+the groups of `gd` under iteration and integer indexing.
+
+# Examples
+
+```jldoctest groupkeys
+julia> df = DataFrame(a = repeat([:foo, :bar, :baz], outer=[4]),
+                      b = repeat([2, 1], outer=[6]),
+                      c = 1:12);
+
+julia> gd = groupby(df, [:a, :b])
+GroupedDataFrame with 6 groups based on keys: a, b
+First Group (2 rows): a = :foo, b = 2
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 2     │ 7     │
+⋮
+Last Group (2 rows): a = :baz, b = 1
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ baz    │ 1     │ 6     │
+│ 2   │ baz    │ 1     │ 12    │
+
+julia> keys(gd)
+6-element DataFrames.GroupKeys{GroupedDataFrame{DataFrame}}:
+ GroupKey: (a = :foo, b = 2)
+ GroupKey: (a = :bar, b = 1)
+ GroupKey: (a = :baz, b = 2)
+ GroupKey: (a = :foo, b = 1)
+ GroupKey: (a = :bar, b = 2)
+ GroupKey: (a = :baz, b = 1)
+```
+
+`GroupKey` objects behave similarly to `NamedTuple`s:
+
+```jldoctest groupkeys
+julia> k = keys(gd)[1]
+GroupKey: (a = :foo, b = 2)
+
+julia> keys(k)
+(:a, :b)
+
+julia> values(k)
+(:foo, 2)
+
+julia> NamedTuple(k)
+(a = :foo, b = 2)
+
+julia> k.a
+:foo
+
+julia> k[:a]
+:foo
+
+julia> k[1]
+:foo
+```
+
+Keys can be used as indices to retrieve the corresponding group from their
+`GroupedDataFrame`:
+
+```jldoctest groupkeys
+julia> gd[k]
+2×3 SubDataFrame
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 2     │ 7     │
+
+julia> isequal(gd[keys(gd)[1]], gd[1])
```

Comment:
Just for simplicity:
```suggestion
julia> gd[keys(gd)[1] == gd[1]
```